synapse/cores/common.py

```diff
@@ -1938,3 +1912,74 @@ def _tufosByInetCidr(self, prop, valu, limit=None):
 
         return self.getTufosBy('range', prop, (ipv4addr, ipv4addr+mask), limit=limit)
 
+    #############################################################
+    # support typelib persistence
+    #############################################################
+
+    # overrides: synapse.lib.types.TypeLib.addType
+    def addType(self, typ):
+        s_types.TypeLib.addType(self, typ)
+        self.formTufoByFrob('syn:type', typ.name)
+
+    # overrides: synapse.lib.types.TypeLib.addSubType
+    def addSubType(self, name, subof, **info):
+        s_types.TypeLib.addSubType(self, name, subof, **info)
+        base = self.getTufoByFrob('syn:type', subof)
+
+        subinfo = {}
+        for propname, propval in base[1].items():
+            if not propname.startswith('syn:type:'):
+                continue
+            propname = propname.partition('syn:type')[2]
+            subinfo[propname] = propval
+
+        for propname, propval in info.items():
+            subinfo[propname] = propval
+
+        subinfo['base'] = subof
+
+        if 'fields' in subinfo:
+            fields = subinfo.pop('fields')
+            subinfo['fields'] = len(fields)
+            for i, (fieldname, fieldtype) in enumerate(fields):
+                k = 'field:%d:%s' % (i, fieldname)
+                subinfo[k] = fieldtype
+
+        # TypeLib.addSubType calls TypeLib.addType, so our subtype tufo already exists,
+        #  but we still need to add the subtype-specific fields
+        sub = self.getTufoByProp('syn:type', name)
+        subinfo = self._frobTufoProps('syn:type', subinfo)
+        self.setTufoProps(sub, **subinfo)
+
+    #############################################################
+    # support datamodel persistence
+    #############################################################
+
+    # overrides: synapse.datamodel.DataModel.addTufoForm
+    def addTufoForm(self, form, **info):
+        DataModel.addTufoForm(self, form, **info)
+        self.formTufoByFrob('syn:form', form, **info)
+
+    # overrides: synapse.datamodel.DataModel.addPropDef
+    def addPropDef(self, prop, **info):
+        DataModel.addPropDef(self, prop, **info)
+
+        form = info.get('form')
+        defval = info.get('defval')
```

Comment:
It would be best to blindly store all elements from info. Maybe we just need to mandate that all type info is cortex compatible. Subfields could be a string ala <name>,<type>[|...]